alicloud/resource_alicloud_oss_bucket_object.go

```diff
@@ -0,0 +1,250 @@
+				ConflictsWith: []string{"content"},
+			},
+
+			"content": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"source"},
+			},
+
+			"acl": {
+				Type:         schema.TypeString,
+				Default:      oss.ACLPrivate,
+				Optional:     true,
+				ValidateFunc: validateOssBucketAcl,
+			},
+
+			"content_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"content_length": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cache_control": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"content_disposition": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"content_encoding": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"content_md5": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"expires": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"server_side_encryption": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateOssBucketObjectServerSideEncryption,
+				Computed:     true,
+			},
+
+			"etag": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAlicloudOssBucketObjectPut(d *schema.ResourceData, meta interface{}) error {
+
+	bucket, err := meta.(*AliyunClient).ossconn.Bucket(d.Get("bucket").(string))
+	if err != nil {
+		return fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	var filePath string
+	var body io.Reader
+
+	if v, ok := d.GetOk("source"); ok {
+		source := v.(string)
+		path, err := homedir.Expand(source)
+		if err != nil {
+			return fmt.Errorf("Error expanding homedir in source (%s): %s", source, err)
+		}
+
+		filePath = path
+	} else if v, ok := d.GetOk("content"); ok {
+		content := v.(string)
+		body = bytes.NewReader([]byte(content))
+	} else {
+		return fmt.Errorf("[ERROR] Must specify \"source\" or \"content\" field")
+	}
+
+	key := d.Get("key").(string)
+	options, err := buildObjectHeaderOptions(d)
+	if err != nil {
+		return err
+	}
+	if filePath != "" {
+		err = bucket.PutObjectFromFile(key, filePath, options...)
+	}
+
+	if body != nil {
+		err = bucket.PutObject(key, body, options...)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error putting object in Oss bucket (%#v): %s", bucket, err)
+	}
+
+	d.SetId(key)
+	return resourceAlicloudOssBucketObjectRead(d, meta)
+}
+
+func resourceAlicloudOssBucketObjectRead(d *schema.ResourceData, meta interface{}) error {
+	bucket, err := meta.(*AliyunClient).ossconn.Bucket(d.Get("bucket").(string))
+	if err != nil {
+		return fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	options, err := buildObjectHeaderOptions(d)
+	if err != nil {
+		return fmt.Errorf("Error building object header options: %#v", err)
+	}
+
+	object, err := bucket.GetObjectDetailedMeta(d.Get("key").(string), options...)
+	if err != nil {
+		return fmt.Errorf("Error Reading Object: %#v", err)
```

Comment:
Is it possible this call will return an `err` that represents a `404 not found`, such that we would return here and not get to check if the `object` is `nil` in the next block, where we would also call `d.SetId("")`?